Pairing Adapters architecture (#450)
kyma-incubator/compass: docs/investigations/pairing-adapters/pairing-with-use-integration-system.md

### Deployment

#### As a sidecar container for the Director component

In this case, the Pairing Adapter container is deployed together with the Director container. This deployment allows the Director to communicate with the given Pairing Adapter through the localhost interface on a given port. The configuration details for the Pairing Adapter needs to be present during the deployment and its lifecycle is bound to the Director component.

The given Integration System stores the internal port on which the Pairing Adapter service listens.

#### As a separate Pod exposed by the Kubernetes service internally (service of the ClusterIP type)

This type of deployment allows to run Pairing Adapter in a separate Pod. The Pairing Adapter runs in the same cluster as the Director component and its API does not have to be publically available. The Pairing Adapter API must be exposed through the Kubernetes service of the ClusterIP type. Having the Kubernetes service, the Director can communicate with the given Pairing Adapter through the DNS name of that service. This approach also allows for using the mTLS communication between components inside the cluster.

> The given Integration System stores the DNS name of the internal Kubernetes service via which the Pairing Adapter is exposed.

Reviewer: ```suggestion
The given Integration System stores the DNS name of the internal Kubernetes service. The Pairing Adapter is exposed through this service.
```

In this case, the Integration System stores the information regarding the credentials for the Pairing Adapter API access together with the required URLs.

### Pairing Adapter contract

To be able to act as a Pairing Adapter, the component must fulfill the following contract:

>**NOTE:** The component offers the GraphQL API. All the queries and mutations work in a synchronous manner.

1. It offers the `getAuthTypes(): [PairingAdapterAuthType!]` query that returns the collection of `PairingAdapterAuthType`.

1. It offers the `generateAuth(type: PairingAdapterAuthType, in: PairingAdapterParamsInput): PairingAuthResult!` mutation that handles the process of obtaining the pairing credentials from the Integration System instance.

`PairingAdapterAuthType` is the type that describes the supported authentication types for the PairingAdapter.

`PairingAdapterParams` is the type that describes the payload which is passed from the Director component to the Pairing Adapter. This payload should be sufficient for Pairing Adapter to obtain pairing data from the Integration System instance. The response from this mutation returns the `PairingAuthResult` type.

> 

Reviewer: ```suggestion

The GraphQL contract for the API looks as follows:

```

# Pairing Application with Integration System

## Introduction

When dealing with multiple Integration Systems each of them may support different ways of establishing trusted relation for the given Application instance. To provide a consistent experience for users, we introduce the concept of Pairing Adapters. Each Integration System can be coupled with one Pairing Adapter which will abstract the details on how the Integration System handles the trusted relation prerequisites.

## Solution overview

This diagram illustrates the flow for pairing the Application with its Integration System Instance using the Pairing Adapter.

> ![Integration System Instance Pairing Adapter](./external-broker-pairing-adapter.svg)

Author: It indicates the steps done by the Application instance. What do you mean by the steps borders?

In this case the Integration System stores the information regarding the credentails for the Pairing Adapter API access together with the required URLs.

### Pairing Adapter contract

The component to be able to act as a Pairing Adapter must fulfill the following contract. 

>**NOTE** The component offers the GraphQL API. All the queries and mutations work in a synchronous manner.

1. It offers the query `getAuthTypes(): [PairingAdapterAuthType!]` that returns collection of `PairingAdapterAuthType`.

1. It offers the mutation `generateAuth(type: PairingAdapterAuthType, in: PairingAdapterParamsInput): PairingAuthResult!` that handles the process of obtaining the pairing credentials from the Integration System Instance.

`PairingAdapterAuthType` is the type that describes supported auth types for the PairingAdapter.

`PairingAdapterParams` is the type that describes the payload which is passed from the Director component to the Pairing Adapter. This payload should be sufficient for Pairing Adapter to obtain pairing data from the Integration System Instance. The response from this mutation returns `PairingAuthResult` type.

> ```graphql

Author: It is the GrapQL contract for the API which is described in this whole paragraph.
